### PR TITLE
[WIP] Orbopt spliner2r impl

### PR DESCRIPTION
**Pull request by jptowns**
## Proposed changes

Preliminary implementation of orbital optimization for splines. Currently only works for real wave functions. Addresses #3451

## What type(s) of changes does this code introduce?

- New feature
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Ubuntu 20 desktop w/ gnu9 compiler and NETLIB

## Checklist

- No. This PR is up to date with current the current state of 'develop'
- No. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WFOpt/QMCCostFunction.cpp [hunk 1]
```diff
@@ -710,8 +721,8 @@ QMCCostFunction::Return_rt QMCCostFunction::fillOverlapHamiltonianMatrices(Matri
     }
   }
 
-  myComm->allreduce(D_avg);
-
+  myComm->allreduce(D_avg);  
+
```
Comment (markdewing) [inline src/QMCDrivers/WFOpt/QMCCostFunction.cpp:725]: There some extra spaces introduced in the PR.  Running clang-format would get rid of them.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp [hunk 2]
```diff
@@ -1264,7 +1261,9 @@ bool QMCFixedSampleLinearOptimize::one_shift_run()
   // build the overlap and hamiltonian matrices
   optTarget->fillOverlapHamiltonianMatrices(hamMat, ovlMat);
   invMat.copy(ovlMat);
-
+  
+  // DEBUG: This is supposed to print the
```
Comment (markdewing) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:1265]: It should be activated with `<parameter name="output_matrices_csv">yes</parameter>` in the optimizer input.  The output goes to files with `.ovl` and '.ham' in the names.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp [hunk 3]
```diff
@@ -1299,11 +1298,12 @@ bool QMCFixedSampleLinearOptimize::one_shift_run()
...
 
   // compute the scaling constant to apply to the update
-  Lambda = getNonLinearRescale(parameterDirections, ovlMat);
-
-  // scale the update by the scaling constant
+  // and then scale the parameter updates
+  Lambda = getNonLinearRescale(parameterDirections, ovlMat);  
   for (int i = 0; i < numParams; i++)
-    parameterDirections.at(i + 1) *= Lambda;
+    {
```
Comment (markdewing) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:1304]: Remove the braces - they are not needed for a single statement for loop body.

File: src/QMCWaveFunctions/RotatedSPOs.h [hunk 4]
```diff
@@ -204,11 +204,21 @@ class RotatedSPOs : public SPOSet
 
   //*********************************************************************************
   //the following functions simply call Phi's corresponding functions
-  void setOrbitalSetSize(int norbs) override { Phi->setOrbitalSetSize(norbs); }
+  void setOrbitalSetSize(int norbs) override
+  {
+    std::cerr << " Entered RotatedSPOs::setOrbitalSetSize()...\n";
```
Comment (markdewing) [inline src/QMCWaveFunctions/RotatedSPOs.h:209]: These look like statements used for debugging, and should be removed.

File: src/QMCWaveFunctions/BsplineFactory/SplineR2R.h [hunk 5]
```diff
@@ -79,6 +88,37 @@ class SplineR2R : public BsplineSet
...
+    return ( SplineInst == nullptr ) ? 0 : SplineInst->num_splines();   
+  }
+
+  // Accumulate the total rotation
+  void storeParamsBeforeRotation() override
+  {
+    std::cerr << "Function SplineR2R::storeParamsBeforeRotation() not implemented!\n";
+  }
+  
+  void applyRotation(const ValueMatrix_t& rot_mat, bool use_stored_copy) override;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/BsplineFactory/SplineR2R.h:119]: Please make a PR with only applyRotation being added and a unit test.
We are missing proper documentation for this function ins `SPOSet.h`.
```
rot_mat
```
should be of size num_occupied x OrbitalSetSize.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp (hunk 2 above)
Comment (jptowns) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimize.cpp:1265]: Fixed.

**Pull request closed without merging**